cpp/src/generators/generator_tools.cu

```diff
@@ -0,0 +1,278 @@
+  thrust::transform(rmm::exec_policy(handle.get_stream()),
+                    pair_first,
+                    pair_first + d_src_v.size(),
+                    pair_first,
+                    [scale] __device__(auto pair) {
+                      return thrust::make_tuple(
+                        experimental::detail::scramble(thrust::get<0>(pair), scale),
+                        experimental::detail::scramble(thrust::get<1>(pair), scale));
+                    });
+}
+
+template <typename vertex_t, typename weight_t>
+std::
+  tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>, rmm::device_uvector<weight_t>>
+  combine_edgelists(raft::handle_t const &handle,
+                    std::vector<rmm::device_uvector<vertex_t>> &&sources,
+                    std::vector<rmm::device_uvector<vertex_t>> &&dests,
+                    std::optional<std::vector<rmm::device_uvector<weight_t>>> &&optional_d_weights)
+{
+  CUGRAPH_EXPECTS(sources.size() == dests.size(),
+                  "sources and dests vertex lists must be the same size");
+
+  if (optional_d_weights) {
+    CUGRAPH_EXPECTS(sources.size() == optional_d_weights.value().size(),
+                    "has_weights is specified, sources and weights must be the same size");
+
+    thrust::for_each_n(
+      thrust::host,
+      thrust::make_zip_iterator(
+        thrust::make_tuple(sources.begin(), dests.begin(), optional_d_weights.value().begin())),
+      sources.size(),
+      [](auto tuple) {
+        CUGRAPH_EXPECTS(thrust::get<0>(tuple).size() != thrust::get<1>(tuple).size(),
+                        "source vertex and dest vertex uvectors must be same size");
+        CUGRAPH_EXPECTS(thrust::get<0>(tuple).size() != thrust::get<2>(tuple).size(),
+                        "source vertex and weights uvectors must be same size");
+      });
+  } else {
+    thrust::for_each_n(
+      thrust::host,
+      thrust::make_zip_iterator(thrust::make_tuple(sources.begin(), dests.begin())),
+      sources.size(),
+      [](auto tuple) {
+        CUGRAPH_EXPECTS(thrust::get<0>(tuple).size() == thrust::get<1>(tuple).size(),
+                        "source vertex and dest vertex uvectors must be same size");
+      });
+  }
+
+  std::vector<rmm::device_uvector<weight_t>> d_weights;
+
+  size_t size{0};
+  for (size_t i = 0; i < sources.size(); ++i) size += sources[i].size();
+
+  rmm::device_uvector<vertex_t> srcs_v(size, handle.get_stream());
+  rmm::device_uvector<vertex_t> dsts_v(size, handle.get_stream());
+  rmm::device_uvector<weight_t> weights_v(0, handle.get_stream());
+
+  detail::append_all<vertex_t>(handle, std::move(sources), srcs_v);
+  detail::append_all<vertex_t>(handle, std::move(dests), dsts_v);
+
+  if (optional_d_weights) {
+    weights_v.resize(size, handle.get_stream());
+    detail::append_all(handle, std::move(optional_d_weights.value()), weights_v);
+  }
+
+  size_t number_of_edges{0};
+
+  if (optional_d_weights) {
```

Comment:
So, should we always remove all multi-edges or better make this as an option? Graph500 input edge lists can possibly have multi-edges (and self-loops) and removing those is a task in the graph construction step.

Reply:
Added a flag to control this behavior in next push.